src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,275 @@
+use super::prelude::*;
+use serde::Serialize;
+
+pub(super) struct HeaderStage<
+    'a,
+    K: AsRef<[u8]>,
+    V: Serialize + 'a,
+    It: Iterator<Item = (K, &'a Vec<V>)> + Clone,
+> {
+    headers_iter: It,
+    header: IndexHeader,
+    key_size: usize,
+    meta: Vec<u8>,
+}
+
+pub(super) struct LeavesStage<
+    'a,
+    K: AsRef<[u8]>,
+    V: Serialize + 'a,
+    It: Iterator<Item = (K, &'a Vec<V>)> + Clone,
+> {
+    headers_iter: It,
+    header: IndexHeader,
+    leaf_size: usize,
+    leaves_buf: Vec<u8>,
+    leaves_offset: u64,
+    headers_size: usize,
+    meta: Vec<u8>,
+}
+
+pub(super) struct TreeStage<
+    'a,
+    K: AsRef<[u8]>,
+    V: Serialize + 'a,
+    It: Iterator<Item = (K, &'a Vec<V>)> + Clone,
+> {
+    headers_iter: It,
+    metadata: TreeMeta,
+    header: IndexHeader,
+    meta_buf: Vec<u8>,
+    tree_buf: Vec<u8>,
+    leaves_buf: Vec<u8>,
+    headers_size: usize,
+    meta: Vec<u8>,
+}
+
+pub(super) struct Serializer<
+    'a,
+    K: AsRef<[u8]>,
+    V: Serialize + 'a,
+    It: Iterator<Item = (K, &'a Vec<V>)> + Clone,
+> {
+    headers_iter: It,
+}
+
+impl<'a, K: AsRef<[u8]> + 'a, V: Serialize + 'a, It: Iterator<Item = (K, &'a Vec<V>)> + Clone>
+    Serializer<'a, K, V, It>
+{
+    pub(super) fn new(headers_iter: It) -> Self {
+        Self { headers_iter }
+    }
+    pub(super) fn header_stage(self, meta: Vec<u8>) -> Result<HeaderStage<'a, K, V, It>> {
+        if let Some((key, Some(record_header))) = self
+            .headers_iter
+            .clone()
```

Comment:
You don't need a `clone()` here, because you can declare `self` as mutable

Reply:
It was needed, because I wanted to extract the first element from iterator, but leave capability to extract it later again (for serialization).
This logic (with generalized serializer) is in #101 now, I removed it from this request, because generalization need to be discussed (and this serializer may affect performance)